### PR TITLE
Subnetworks API implementation

### PR DESCRIPTION
**Pull request by olperr1**
**Please check if the PR fulfills these requirements**

- [X] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem?**

No

**What kind of change does this PR introduce?**

Feature

**What is the current behavior?**

The subnetwork API is not implemented.

**What is the new behavior (if this is a feature change)?**
Implements the subnetworks API

**Does this PR introduce a breaking change or deprecate an API?**
No

### CODE REVIEW

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/SubnetworkImpl.java [hunk 1]
```diff
@@ -686,8 +695,43 @@ public void merge(Network... others) {
...
     public Network detach() {
-        //TODO subnetworks API
-        throw new UnsupportedOperationException("Not yet implemented");
+        Set<Identifiable<?>> boundaryElements = getBoundaryElements();
+        checkDetachable(boundaryElements, true);
+
+        long start = System.currentTimeMillis();
+
+        // Remove tie-lines
+        boundaryElements.stream()
```
Comment (annetill) [inline iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/SubnetworkImpl.java:704]: P0 and Q0 of the dangling lines must be updated.

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/SubnetworkImpl.java [hunk 2]
```diff
@@ -696,20 +741,103 @@ public Network detach() {
...
+    private boolean checkDetachable(Set<Identifiable<?>> boundaryElements, boolean throwsException) {
+        if (parent.getVariantManager().getVariantArraySize() != 1) {
+            if (throwsException) {
+                throw new PowsyblException("Detaching from multi-variants network is not supported");
+            }
+            return false;
+        }
+        if (boundaryElements.stream().anyMatch(Predicate.not(SubnetworkImpl::isSplittable))) {
+            if (throwsException) {
+                throw new PowsyblException("Some un-splittable equipments prevent the subnetwork to be detached");
```
Comment (annetill) [inline iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/SubnetworkImpl.java:756]: typo: boundary elements

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/SubnetworkImpl.java [hunk 3]
```diff
@@ -696,20 +741,103 @@ public Network detach() {
...
+
+    /**
+     * Return all the potential boundary elements: elements defined in the current subnetwork or in the parent network
+     * and which type corresponds to a "linking" element (branch, transformer, ...).
+     *
+     * @return a {@link Stream} of the potential boundary elements
+     */
+    private Stream<Identifiable<?>> getPotentialBoundaryElements() {
+        Stream<Line> lines = parent.getLineStream();
+        Stream<TwoWindingsTransformer> twoWindingsTransformerStream = parent.getTwoWindingsTransformerStream();
```
Comment (annetill) [inline iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/SubnetworkImpl.java:782]: To check: two and three windings transformers cannot be outside of a substation. A substation cannot be shared into two subnetworks.

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/SubnetworkImpl.java (hunk 1 above)
Comment (olperr1) [reply iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/SubnetworkImpl.java:705]: This will be done in the subnetworks main PR.

**Review by annetill — APPROVED** [review]
Do not forget to update `P0` and `Q0` during tie line removal.

**Pull request merged**